### ISSUE TITLE
Symlinks are broken in SFTP

### ISSUE THREAD

**Issue by phawxby** [icetee/remote-ftp#279]
**Atom Version**: 1.4.3
**System**: Microsoft Windows 10 Enterprise
**Thrown From**: [Remote-FTP](https://github.com/mgrenier/remote-ftp) package, v0.7.16
### Stack Trace

Uncaught TypeError: Cannot read property 'isFile' of undefined

```
At C:\Users\z269cjbb\.atom\packages\Remote-FTP\lib\connectors\sftp.js:155

TypeError: Cannot read property 'isFile' of undefined
    at C:\Users\z269cjbb\.atom\packages\Remote-FTP\lib\connectors\sftp.js:155:21
    at SFTPStream._transform (C:\Users\z269cjbb\.atom\packages\Remote-FTP\node_modules\ssh2\node_modules\ssh2-streams\lib\sftp.js:388:17)
    at SFTPStream.Transform._read (_stream_transform.js:167:10)
    at SFTPStream._read (C:\Users\z269cjbb\.atom\packages\Remote-FTP\node_modules\ssh2\node_modules\ssh2-streams\lib\sftp.js:173:15)
    at SFTPStream.Transform._write (_stream_transform.js:155:12)
    at doWrite (_stream_writable.js:292:12)
    at writeOrBuffer (_stream_writable.js:278:5)
    at SFTPStream.Writable.write (_stream_writable.js:207:11)
    at Channel.ondata (_stream_readable.js:525:20)
    at emitOne (events.js:77:13)
    at Channel.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at Channel.Readable.push (_stream_readable.js:110:10)
    at SSH2Stream.<anonymous> (C:\Users\z269cjbb\.atom\packages\Remote-FTP\node_modules\ssh2\lib\Channel.js:140:15)
    at emitOne (events.js:77:13)
    at SSH2Stream.emit (events.js:169:7)
    at parsePacket (C:\Users\z269cjbb\.atom\packages\Remote-FTP\node_modules\ssh2\node_modules\ssh2-streams\lib\ssh.js:3163:10)
    at SSH2Stream._transform (C:\Users\z269cjbb\.atom\packages\Remote-FTP\node_modules\ssh2\node_modules\ssh2-streams\lib\ssh.js:552:13)
    at SSH2Stream.Transform._read (_stream_transform.js:167:10)
    at SSH2Stream._read (C:\Users\z269cjbb\.atom\packages\Remote-FTP\node_modules\ssh2\node_modules\ssh2-streams\lib\ssh.js:213:15)
    at SSH2Stream.Transform._write (_stream_transform.js:155:12)
    at doWrite (_stream_writable.js:292:12)
    at writeOrBuffer (_stream_writable.js:278:5)
    at SSH2Stream.Writable.write (_stream_writable.js:207:11)
    at Socket.ondata (_stream_readable.js:525:20)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at Socket.Readable.push (_stream_readable.js:110:10)
    at TCP.onread (net.js:523:20)
```
### Commands

```
```
### Config

``` json
{
  "core": {}
}
```
### Installed Packages

``` coffee
# User
minimap, v4.19.0
open-recent, v5.0.0
pigments, v0.23.1
Remote-FTP, v0.7.16

# Dev
No dev packages
```

**Comment by jpxd** [comment]
Yeah... my commit broke something. >.< Looking into it right now

**Comment by jpxd** [comment]
Can you checkout the `test-symlinks` branch, install and link it, and post the developer console output here? 

```
git clone https://github.com/mgrenier/remote-ftp.git
cd remote-ftp
git checkout test-symlinks
apm install
apm link
```

**Comment by nlanglo** [comment]
Do you plan to revert to a working version of the plugin, or wait for this guy to debug it for you? Atom plugin remote-ftp is broken for me too.

**Comment by ghost** [comment]
I checked out the branch you mentioned, but got the same error as before.
[Enter steps to reproduce below:]
1. The error occurs when I try to expand a symlink to an other folder.

**Atom Version**: 1.4.3
**System**: Unknown Windows Version
**Thrown From**: [Remote-FTP](https://github.com/mgrenier/remote-ftp) package, v0.7.16

### Stack Trace

Uncaught TypeError: Cannot read property 'isFile' of undefined

```
At C:\Users\...\.atom\packages\remote-ftp\lib\connectors\sftp.js:158

TypeError: Cannot read property 'isFile' of undefined
    at C:\Users\...\.atom\packages\remote-ftp\lib\connectors\sftp.js:158:21
    at SFTPStream._transform (C:\Users\...\.atom\packages\remote-ftp\node_modules\ssh2\node_modules\ssh2-streams\lib\sftp.js:388:17)
    at SFTPStream.Transform._read (_stream_transform.js:167:10)
    at SFTPStream._read (C:\Users\...\.atom\packages\remote-ftp\node_modules\ssh2\node_modules\ssh2-streams\lib\sftp.js:173:15)
    at SFTPStream.Transform._write (_stream_transform.js:155:12)
    at doWrite (_stream_writable.js:292:12)
    at writeOrBuffer (_stream_writable.js:278:5)
    at SFTPStream.Writable.write (_stream_writable.js:207:11)
    at Channel.ondata (_stream_readable.js:525:20)
    at emitOne (events.js:77:13)
    at Channel.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at Channel.Readable.push (_stream_readable.js:110:10)
    at SSH2Stream.<anonymous> (C:\Users\...\.atom\packages\remote-ftp\node_modules\ssh2\lib\Channel.js:140:15)
    at emitOne (events.js:77:13)
    at SSH2Stream.emit (events.js:169:7)
    at parsePacket (C:\Users\...\.atom\packages\remote-ftp\node_modules\ssh2\node_modules\ssh2-streams\lib\ssh.js:3163:10)
    at SSH2Stream._transform (C:\Users\...\.atom\packages\remote-ftp\node_modules\ssh2\node_modules\ssh2-streams\lib\ssh.js:552:13)
    at SSH2Stream.Transform._read (_stream_transform.js:167:10)
    at SSH2Stream._read (C:\Users\...\.atom\packages\remote-ftp\node_modules\ssh2\node_modules\ssh2-streams\lib\ssh.js:213:15)
    at SSH2Stream.Transform._write (_stream_transform.js:155:12)
    at doWrite (_stream_writable.js:292:12)
    at writeOrBuffer (_stream_writable.js:278:5)
    at SSH2Stream.Writable.write (_stream_writable.js:207:11)
    at Socket.ondata (_stream_readable.js:525:20)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at Socket.Readable.push (_stream_readable.js:110:10)
    at TCP.onread (net.js:523:20)
```

### Commands

```
     -1:06.8.0 core:confirm (atom-text-editor.editor.mini.is-focused)
```

### Config

``` json
{
  "core": {
    "disabledPackages": [
      "remote-edit",
      "git-wip",
      "vim-mode"
    ],
    "themes": [
      "atom-light-ui",
      "base16-tomorrow-light-theme"
    ]
  }
}
```

### Installed Packages

``` coffee
# User
atom-alignment, v0.12.1
change-case, v0.6.0
color-picker, v2.1.1
emmet, v2.4.1
highlight-selected, v0.11.2
jshint, v1.8.3
language-lua, v0.9.4
pigments, v0.23.1

# Dev
No dev packages
```

**Comment by jpxd** [comment]
@gtv-p-vanpoucke, @phawxby  The bug is not fixed in the `test-symlinks` branch, I just added logging so I can understand whats going wrong on your end. Please open the developer console, reproduce the error and post the output here.

I will revert the commit now in the main branch but I need your help to fix this issue, because on my end everything is working fine with SFTP symlink files & directories. Prior to this commit all symlinks were shown as files, which was also bad.

**Comment by czmole** [comment]
On my end, prior to the commit, I was using a symlink as a default path for the SFTP connection but the other symlinks iniside that were not shown as folders.

**Comment by ghost** [comment]
This is what I get in the console window.  Hope it helps...

[remoteftp-1455010635433.201602091049.log.txt](https://github.com/mgrenier/remote-ftp/files/122843/remoteftp-1455010635433.201602091049.log.txt)
[remoteftp-1455010635433.stripped.log.txt](https://github.com/mgrenier/remote-ftp/files/122844/remoteftp-1455010635433.stripped.log.txt)

**Comment by jpxd** [comment]
@gtv-p-vanpoucke thx!
Just one question: does `/home/user/Documenten/Projecten/XTransport/Xtransport/htdocs/settings/XTransWebservice/lib`
exist?

Because the listing fails at resolving this symlink

```
/home/user/Documenten/Projecten/XTransport/Xtransport/htdocs/XScripts/lib -> /home/user/Documenten/Projecten/XTransport/Xtransport/htdocs/settings/XTransWebservice/lib
err Error: No such file(…)
```

If it is just a "broken" symlink, we could catch that exception and show a user-friendly warning about it.

**Comment by ghost** [comment]
@jpxd it doesn't, but `/home/user/Documenten/Projecten/XTransport/Xtransport/settings/XTransWebservice/lib` does.  The symlink `XScripts` refers to `../Xscripts/`.  The symlink `lib` in `../Xscripts/` refers to `../settings/XTransWebservice/lib`.

**Comment by jpxd** [comment]
@gtv-p-vanpoucke So do I understand correctly, the (simplified) chain is as follows?

```
XTransport/Xtransport/htdocs/XScripts --> ../Xscripts/
XTransport/Xtransport/XScripts/lib --> ../settings/XTransWebservice/lib
XTransport/Xtransport/settings/XTransWebservice/lib
```

Edit: Can reproduce the error now

**Comment by ghost** [comment]
@jpxd The chain is correct.

**Comment by jpxd** [comment]
Should work in the new release :smiley:

**Comment by nlanglo** [comment]
should work, but doesn't, it's broken again

**Comment by jpxd** [comment]
Which version are you using? Can you give me some information about your symlink structure?

**Comment by nlanglo** [comment]
atom 1.5.2
remote-ftp 0.7.18
same stack trace error as before

**Comment by chrismerat** [comment]
I receive an error for a symlink on connecting and noticed that the paths are incorrect. It displays the error as `/home/site-name/public_html/flyers -> /home/site-name/public_html/home/other-site-name/public_html/flyers` which is including the root path of the SFTP connection at the beginning of the symlinked resource.

The symlink is structured as `flyers -> /home/other-site-name/public_html/flyers` on the server when viewing from the `/home/site-name/public_html` directory.

**Comment by jpxd** [comment]
Should be fixed in the latest release (v0.8.1), but try for yourself....
Now `realpath` is used in order to resolve symlinks, which should handle relative symlinks in the right way.

**Comment by nlanglo** [comment]
0.8.1 fixed the issue for me, no more error message either, thanks.
